examples/src/main/java/com/mapbox/navigation/examples/core/ReplayEngineActivity.kt

```diff
@@ -0,0 +1,266 @@
+
+        val mapboxNavigationOptions = MapboxNavigation.defaultNavigationOptions(
+                this,
+                Utils.getMapboxAccessToken(this)
+        )
+
+        mapboxNavigation = MapboxNavigation(
+                applicationContext,
+                Utils.getMapboxAccessToken(this),
+                mapboxNavigationOptions,
+                locationEngine = replayRouteLocationEngine
+        )
+
+        initListeners()
+    }
+
+    override fun onMapReady(mapboxMap: MapboxMap) {
+        mapboxMap.setStyle(Style.MAPBOX_STREETS) { style ->
+            locationEngine = LocationEngineProvider.getBestLocationEngine(this)
+            initLocationComponent(style, mapboxMap)
+            navigationMapboxMap = NavigationMapboxMap(mapView, mapboxMap).also {
+                it.addProgressChangeListener(mapboxNavigation!!)
+                mapInstanceState?.let { state ->
+                    it.restoreFrom(state)
+                }
+            }
+        }
+        mapboxMap.addOnMapLongClickListener { latLng ->
+            mapboxMap.locationComponent.lastKnownLocation?.let { originLocation ->
+                mapboxNavigation?.requestRoutes(
+                        RouteOptions.builder().applyDefaultParams()
+                                .accessToken(Utils.getMapboxAccessToken(applicationContext))
+                                .coordinates(originLocation.toPoint(), null, latLng.toPoint())
+                                .alternatives(true)
+                                .profile(DirectionsCriteria.PROFILE_DRIVING_TRAFFIC)
+                                .build(),
+                        routesReqCallback
+                )
+            }
+            true
+        }
+        locationComponent = mapboxMap.locationComponent
+    }
+
+    @SuppressLint("RestrictedApi")
+    fun initLocationComponent(loadedMapStyle: Style, mapboxMap: MapboxMap) {
+        when (PermissionsManager.areLocationPermissionsGranted(this)) {
+            true -> {
+                mapboxMap.moveCamera(CameraUpdateFactory.zoomTo(15.0))
+                mapboxMap.locationComponent.let { locationComponent ->
+                    val locationComponentActivationOptions =
+                            LocationComponentActivationOptions.builder(this, loadedMapStyle)
+                                    // .useDefaultLocationEngine(true)
+                                    .build()
+
+                    locationComponent.activateLocationComponent(locationComponentActivationOptions)
+                    locationComponent.isLocationComponentEnabled = true
+                    locationComponent.cameraMode = CameraMode.TRACKING
+                    locationComponent.renderMode = RenderMode.COMPASS
+
+                    initLocationEngine()
+                }
+            }
+            false -> {
+                permissionsManager =
+                        PermissionsManager(this).also { it.requestLocationPermissions(this) }
```

Comment:
since we are checking permission here, would it be better to implement the permission request callbacks below? https://github.com/mapbox/mapbox-navigation-android/pull/2611/files#diff-de2cfae3c1e914451e4457f7979cb8a4R220-R233
Otherwise, maybe better to not check the permission.

Reply:
Abhishek and I discussed checking the permissions and we thought it was appropriate since it would be more complete as an example.  As far as what to do when the user allows or denies the permission that seems outside the scope of the example. Requesting permission would be needed to run the example but handling the response is dependent on the the developer's use case.

Reply:
right, I am not against permission check. As we agree that it would be more complete as an example for having it. I am suggesting to implement the `onPermissionResult` callback as well. Otherwise, it looks like incomplete.
Let's say a developer just copy paste this code into their small project, nothing will happen after the user grants the permission.
If you two agree on this, I am fine with it.

Reply:
and what I am suggesting looks like:
```java
    override fun onCreate(savedInstanceState: Bundle?) {
        ...

        if (hasPermission) {
            initLocationComponent()
        } else {
            requestPermission()
        }

        ...
    }

    override fun onPermissionResult(granted: Boolean) {
        when (granted) {
            true -> {
                initLocationComponent()
            }
            false -> {
                finish()
            }
        }
    }
```
it will be a more complete example IMO.

Reply:
I agree that we should handle the permissions completely thereby giving a full example or if we think that this is too much of extra code that we as SDK developer shouldn't think much about in these examples, we can always tell the external developer to make sure they allow location permissions for proper functioning of the SDK.

Reply:
Trying to address all the permission use cases and on a 2nd thought I feel that we should not worry about asking permissions everytime in each Activity. The main idea of writing these Activities is to show them how to use our SDK components. Adding permissions will involve in a lot of these `hasPermissions` check in the code thereby making it unreadable. I think we should keep our code concise and handle location permissions in the `MainActivity` as we are doing. Let's not enforce them directly but show them appropriate message.
Wdyt @cafesilencio @olegzil @Guardiola31337 @JunDai

Reply:
> I think we should keep our code concise and handle location permissions in the MainActivity as we are doing

I like this 'cause it'd keep the examples cleaner. This is done in a number of other Mapbox SDK test apps and it works well. One thing you could do in the kdocs for each example activity is to note that location permission checking is happening in `MainActivity`.

Reply:
I think the MainActivity shows the permission handle is enough so we don't need to worry about it in each example activity.
In the other hand, my point is if we do have the request permission logic in the example activity, then we should do it completely by implementing the permission request callbacks.